Dockerfile

```diff
@@ -1,24 +1,144 @@
+
+# runs golangci-lint
+FROM base AS lint-golangci-lint
+WORKDIR /src
+COPY .golangci.yml .
+ENV GOGC 50
+RUN golangci-lint config verify --config .golangci.yml
+RUN --mount=type=cache,target=/root/.cache/go-build --mount=type=cache,target=/root/.cache/golangci-lint --mount=type=cache,target=/go/pkg golangci-lint run --config .golangci.yml
+
+# runs govulncheck
+FROM base AS lint-govulncheck
+WORKDIR /src
+RUN --mount=type=cache,target=/root/.cache/go-build --mount=type=cache,target=/go/pkg govulncheck ./...
+
+# runs unit-tests with race detector
+FROM base AS unit-tests-race
+WORKDIR /src
+ARG TESTPKGS
+RUN --mount=type=cache,target=/root/.cache/go-build --mount=type=cache,target=/go/pkg --mount=type=cache,target=/tmp CGO_ENABLED=1 go test -v -race -count 1 ${TESTPKGS}
+
+# runs unit-tests
+FROM base AS unit-tests-run
+WORKDIR /src
+ARG TESTPKGS
+RUN --mount=type=cache,target=/root/.cache/go-build --mount=type=cache,target=/go/pkg --mount=type=cache,target=/tmp go test -v -covermode=atomic -coverprofile=coverage.txt -coverpkg=${TESTPKGS} -count 1 ${TESTPKGS}
+
+FROM embed-generate AS embed-abbrev-generate
+WORKDIR /src
+ARG ABBREV_TAG
+RUN echo -n 'undefined' > internal/version/data/sha && \
+    echo -n ${ABBREV_TAG} > internal/version/data/tag
+
+FROM scratch AS unit-tests
+COPY --from=unit-tests-run /src/coverage.txt /coverage-unit-tests.txt
+
+# cleaned up specs and compiled versions
+FROM scratch AS generate
+COPY --from=embed-abbrev-generate /src/internal/version internal/version
+
+# builds talos-vmtoolsd-linux-amd64
+FROM base AS talos-vmtoolsd-linux-amd64-build
+COPY --from=generate / /
+COPY --from=embed-generate / /
+WORKDIR /src/cmd/talos-vmtoolsd
+ARG GO_BUILDFLAGS
+ARG GO_LDFLAGS
+ARG VERSION_PKG="internal/version"
+ARG SHA
+ARG TAG
+RUN --mount=type=cache,target=/root/.cache/go-build --mount=type=cache,target=/go/pkg go build ${GO_BUILDFLAGS} -ldflags "${GO_LDFLAGS} -X ${VERSION_PKG}.Name=talos-vmtoolsd -X ${VERSION_PKG}.SHA=${SHA} -X ${VERSION_PKG}.Tag=${TAG}" -o /talos-vmtoolsd-linux-amd64
+
+FROM scratch AS talos-vmtoolsd-linux-amd64
+COPY --from=talos-vmtoolsd-linux-amd64-build /talos-vmtoolsd-linux-amd64 /talos-vmtoolsd-linux-amd64
+
+FROM talos-vmtoolsd-linux-${TARGETARCH} AS talos-vmtoolsd
+
+FROM scratch AS talos-vmtoolsd-all
+COPY --from=talos-vmtoolsd-linux-amd64 / /
+
+FROM scratch AS image-talos-vmtoolsd
 ARG TARGETARCH
-ARG CGO_ENABLED=0
-ARG GOARCH=$TARGETARCH
-ARG GOOS=linux
-RUN go test -v ./... && \
-    go vet ./... && \
-    go build -ldflags="-s -w" -trimpath -o talos-vmtoolsd ./cmd/talos-vmtoolsd
-
-#### Build system extension tree
-FROM alpine:3.19 AS stage
-RUN mkdir -p /stage/rootfs/usr/local/etc/containers
-RUN mkdir -p /stage/rootfs/usr/local/lib/containers/talos-vmtoolsd
-COPY --from=builder /build/talos-vmtoolsd /stage/rootfs/usr/local/lib/containers/talos-vmtoolsd/
-COPY ./manifest.yaml /stage/
-COPY ./talos-vmtoolsd.yaml /stage/rootfs/usr/local/etc/containers/
```

Comment:
Doesn't this remove the ability to use this container as system extension?

Reply:
Ahh I didn't realize this was supposed to be used like that, I though this would go as an official system extension? :thinking:

Reply:
I'm not sure, let's ask @jonkerj what was agreed when this code was moved to `siderolabs/`.

Reply:
cool, else I can add this back, kres should support it

Reply:
Okay, thanks. Can you leave the last lines in as well, including `ENTRYPOINT`? This way the container is usable as a `Pod` as well as a system extension.

I believe the idea was have this repository for the extension short term at least, and _maybe_ add it to siderolabs/extensions later.

Reply:
> I'm not sure, let's ask @jonkerj what was agreed when this code was moved to `siderolabs/`.

Yes, the ambition was to have `talos-vmtoolsd` as an official system extension at some point. No promises were made, but we'd like to put in the effort to get there.

Reply:
Okay, I've resolved this